idx/bigtable/bigtable.go

```diff
@@ -0,0 +1,557 @@
+		attempts := 0
+
+		for !success {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+			errs, err := b.tbl.ApplyBulk(ctx, rowKeys, mutations)
+			cancel()
+			if err != nil {
+				log.Errorf("bigtable-idx: unable to apply writes to bigtable. %s", err)
+				statQueryInsertFail.Add(len(rowKeys))
+				if (attempts % 20) == 0 {
+					log.Warnf("bigtable-idx: Failed to write %d defs to bigtable. they will be retried. %s", len(rowKeys), err)
+				}
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else if len(errs) > 0 {
+				failedRowKeys := make([]string, 0)
+				failedMutations := make([]*bigtable.Mutation, 0)
+				for i, err := range errs {
+					if err != nil {
+						failedRowKeys = append(failedRowKeys, rowKeys[i])
+						failedMutations = append(failedMutations, mutations[i])
+					}
+				}
+				rowKeys = failedRowKeys
+				mutations = failedMutations
+				log.Errorf("bigtable-idx: failed to write %d rows. %s", len(failedRowKeys), err)
+				statQueryInsertFail.Add(len(failedRowKeys))
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else {
+				success = true
+				statQueryInsertExecDuration.Value(time.Since(pre))
+				statQueryInsertOk.Add(len(rowKeys))
+				log.Debugf("bigtable-idx: %d metricDefs saved to bigtable.", len(rowKeys))
+			}
+
+		}
+		buffer = buffer[:0]
+	}
+
+LOOP:
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(time.Second)
+			if len(buffer) > 0 {
+				flush()
+			}
+		case req, ok := <-b.writeQueue:
+			if !ok {
+				break LOOP
+			}
+			buffer = append(buffer, req)
+			if len(buffer) >= writeMaxFlushSize {
+				// make sure the timer hasn't already fired. If it has we read
+				// from the chan and consume the event.
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(time.Second)
+				flush()
+			}
+		}
+	}
+	log.Info("bigtable-idx: writeQueue handler ended.")
+	b.wg.Done()
+}
+
+func (b *BigtableIdx) Delete(orgId uint32, pattern string) ([]idx.Archive, error) {
+	pre := time.Now()
+	defs, err := b.MemoryIdx.Delete(orgId, pattern)
+	if err != nil {
+		return defs, err
+	}
+	if updateBigTableIdx {
+		for _, def := range defs {
+			err = b.deleteDef(&def.MetricDefinition)
+			if err != nil {
+				break
+			}
+		}
+	}
+	statDeleteDuration.Value(time.Since(pre))
+	return defs, err
+}
+
+func (b *BigtableIdx) deleteDef(def *schema.MetricDefinition) error {
+	return b.deleteRow(FormatRowKey(def.Id, def.Partition))
+}
+
+func (b *BigtableIdx) deleteRow(key string) error {
+	pre := time.Now()
+	mut := bigtable.NewMutation()
+	mut.DeleteRow()
+	err := b.tbl.Apply(context.Background(), key, mut)
+	if err != nil {
+		statQueryDeleteFail.Inc()
+		log.Errorf("bigtable-idx: Failed to delete row %s from bigtable. %s", key, err)
+		return err
+	}
```

Comment:
any particular reason why there's no retry logic here, unlike the cassandra counterpart and unlike elsewhere in this plugin?

Reply:
nice